proposals/0000-unsaturated-type-families.rst

```diff
@@ -0,0 +1,1051 @@
+
+  -- inferred: Functor :: (Type -> @M Type) -> @M Constraint
+  class Functor (f :: Type -> Type) where
+
+Arrow-type in type class instances
+##################################
+
+When defining an instance, the arrow type can turn up directly
+in the instance head, for example::
+
+  instance Monad ((->) r)
+  instance Category (->)
+  instance Semigroup (a -> b)
+
+To retain compability, all of these arrows are assumed to mean the term-level
+arrow, in other words unmatchable. This default can be overridden ::
+
+  instance Foo ((->) @M)
+
+Application in pattern position
+###############################
+
+When a kind arrow arises from an application in either a type family pattern or
+a type class instance head, the arrows must be matchable as a *hard rule*  ::
+
+  type family UnApp a where
+    UnApp (f x) = x
+
+  instance Show (g b)
+
+``f`` and ``g`` are inferred to have a matchable kind. Indeed, they must have a matchable
+kind, and declaring otherwise is an error.
+
+RHS of type synonyms
+####################
+
+When writing::
+
+  type Arrow = (->)
+
+the arrow is *defaulted* to mean ``(->) @U``.
+
+Note that making either choice here is a breaking change.
+For example, today one can write ::
+
+  data Maybe :: Arrow Type Type where ...
+
+but this will no longer typecheck because the arrow means unmatchable. The
+decision to default to matchable in this case is grounded in the observation
+that most such synonyms today refer to term-level, thus unmatchable arrows.
+
+A notable exception is the defunctionalisation arrow from the
+`singletons <https://hackage.haskell.org/package/singletons-2.7/docs/Data-Singletons.html#t:-126--62->`_ library::
+
+  type (~>) a b = TyFun a b -> Type
+
+which really refers to a kind-level matchable arrow. However, we expect many
+such use cases to be subsumed by first class higher-order functions introduced
+by this proposal.
+
+Arity of type families
+~~~~~~~~~~~~~~~~~~~~~~
+
+A technical consequence of the proposed framework is that the `arity
+<https://downloads.haskell.org/~ghc/8.10.2/docs/html/users_guide/glasgow_exts.html?highlight=typefamilies#type-family-declarations>`_
+of a type family can not be directly deduced from its kind (although even before
+this proposal that was already the case).
+
+Consider the following two type families ::
+
+  type family Foo (a :: Type) :: Type
+  type family Bar :: Type -> @U Type
+
+Both have the same kind, namely ``Type -> @U Type``, but  the arity of ``Foo``
+is 1, whereas ``Bar`` is nullary. Since partial application is now possible, the
+arities no longer play such an important role. The main place where they still
+show up is in the definitions of type families. Type family equations must bind
+all of their arguments on the left-hand side ::
+
+  type family Foo (a :: Type) :: Type where
+    Foo Int  = Bool
+    Foo Char = Int
+
+but ``Bar``, a nullary type family, can only be defined without arguments and a
+type family on its RHS ::
```

Comment:
Perhaps this section should be moved to the effects & interactions? Its primary aim is to highlight an already existing behaviour that is not changed by this proposal. The reason I felt it important to highlight is because I think it's a valid question to ask whether this is changed now that we can partially apply type families. But maybe it isn't? After re-reading it several times, I'm starting to think this section is not as important as I had thought.

Reply:
> The reason I felt it important to highlight is because I think it's a valid question to ask whether this is changed now that we can partially apply type families.

Yeah, I think it would make sense to move this section to "Effects & Interactions" then. You can say "you might wonder whether this existing restriction can be dropped, alas no because ..."

Reply:
Done, thanks!